File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.cpp
#include <string>
#include <vector>
#include <queue>
#include <iostream>

using namespace std;

queue<pair<int,int>> q;

int solution(int bridge, int weight, vector<int> trucks) {
    int idx = 0;
    int cur = 0;
    int sum = 0;
    
    while(true){
        // 탈출조건
        
        if(trucks.size() <= idx && q.empty()) break;
        cur++;
        // 큐가 비지않았으면,
        while(!q.empty()){
            int w = q.front().first;
            int prev = q.front().second;
            if(cur - prev < bridge) break;
            q.pop();
            sum -= w;
        }
        
        if(sum + trucks[idx] > weight || trucks.size() <= idx) continue; 
    
        q.push({trucks[idx], cur});
        sum += trucks[idx];
        idx++;
    }
              
    return cur;
}

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.io.*;
import java.util.*;

class Pair {
    public int x;
    public int y;
    
    Pair(int x, int y){
        this.x = x;
        this.y = y;
    }
}

class Solution {
    public int solution(int l, int w, int[] tw) {
        int answer = 0;
        Queue<Pair> passing = new ArrayDeque<>();
        
        int n = tw.length;
        int t = 0;
        int cur = 0;
        int sum = 0;

        while(true){
           if(passing.isEmpty() && cur == n) break;
           if(!passing.isEmpty() && passing.peek().y >= l) {
               Pair out = passing.poll();
               sum -= out.x;
           }
         
           if(cur < n && sum + tw[cur] <= w) {
               Pair in = new Pair(tw[cur], 0);
               sum += in.x;
               passing.add(in);
               cur++;
           }
          
           for(Pair p : passing){
               p.y++;
           }
           t++;
        }
        
        return t;
    }
}